File: Programming-Abstraction/Binary-Search_Tree-Java/src/BinarySearchTree.java
import java.util.*;
import java.util.stream.*;

public class BinarySearchTree<T extends Comparable<T>> implements Iterable<T> {
    private BinarySearchTreeNode<T> root;
    private BinarySearchTreeNode<T> cursor;

    private int size;
    private String name;
    static boolean done = false;

    public BinarySearchTree(String name) {
        this.root = null;
        this.cursor = root;
        this.size = 0;
        this.name = name;
    }

    public BinarySearchTreeNode<T> getCursor() {
        return cursor;
    }

    public void CursorToRoot() {
        cursor = root;
    }

    public void cursorToLeft() {
        this.cursor = cursor.getLeft();
    }

    public void cursorToRight() {
        this.cursor = cursor.getRight();
    }

    public BinarySearchTreeNode<T> getRoot() {
        return root;
    }

    public void setRoot(BinarySearchTreeNode<T> root) {
        this.root = root;
    }

    public void setCursor(BinarySearchTreeNode<T> cursor) {
        this.cursor = cursor;
    }

    public void addAll(List<T> asList) {
        for (T i : asList) {
            ++size;
            add(new BinarySearchTreeNode<T>(i));
        }
    }

    @Override
    public Iterator<T> iterator() {
        return treeAsList(root, new ArrayList<T>()).stream().sorted().collect(Collectors.toList()).iterator();
    }

    public void add(BinarySearchTreeNode<T> newNode) {

        if (root == null) {

            root = newNode;
            cursor = root;

        } else {
            if (cursor.getName().compareTo(newNode.getName()) > 0) {
                if (cursor.getLeft() == null) {
                    cursor.setLeft(newNode);
                    CursorToRoot();
                } else {
                    cursorToLeft();
                    add(newNode);
                }

            }
            if (cursor.getName().compareTo(newNode.getName()) < 0) {
                if (cursor.getRight() == null) {
                    cursor.setRight(newNode);
                    CursorToRoot();
                } else {
                    cursorToRight();
                    add(newNode);
                }

            }

        }

    }

    public List<T> treeAsList(BinarySearchTreeNode<T> tree, List<T> root) {
        root.add(tree.getName());
        if (tree.getLeft() != null) {
            treeAsList(tree.getLeft(), root);
        }
        if (tree.getRight() != null) {
            treeAsList(tree.getRight(), root);
        }
        return root;

    }

    public static <T extends Comparable<T>> List<T> merge(BinarySearchTree<T> t1, BinarySearchTree<T> t2) {
        ArrayList<T> mergedList = new ArrayList<T>();
        Object lock = new Object();

        Thread theThreadOne = new Thread(new Runnable() {
            Iterator<? extends T> lst1 = t1.iterator();

            @Override
            public void run() {

                while (lst1.hasNext()) {
                    synchronized (lock) {
                        lock.notifyAll();
                        if (!done) {
                            try {
                                lock.wait();
                            } catch (InterruptedException e) {
                                e.printStackTrace();
                            }
                        }

                        mergedList.add(lst1.next());
                        lock.notifyAll();
                    }
                }
                done = true;

            }
        });
        Thread theThreadTwo = new Thread(new Runnable() {
            Iterator<? extends T> lst2 = t2.iterator();

            @Override
            public void run() {

                while (lst2.hasNext()) {
                    synchronized (lock) {
                        lock.notifyAll();
                        if (!done) {
                            try {
                                lock.wait();
                            } catch (InterruptedException e) {
                                e.printStackTrace();
                            }
                        }

                        mergedList.add(lst2.next());
                        lock.notifyAll();
                    }
                }

            }

        });
        theThreadOne.start();
        theThreadTwo.start();
        try {
            theThreadOne.join();
            theThreadTwo.join();
        } catch (InterruptedException e) {
            // TODO Auto-generated catch block
            e.printStackTrace();
        }
        BinarySearchTree<T> ansTree = new BinarySearchTree<>("");
        ansTree.addAll(mergedList);
        return mergedList.stream().sorted().collect(Collectors.toList());

    }

    public String toString() {
        String tmp = String.format("[%s] %s", name, root.getName());
        List<String> ans = toStringHelper(root, new ArrayList<String>(), tmp, new Stack<String>());
        String strAns = "";
        for (String i : ans) {
            strAns += i;
        }
        return strAns;
    }

    public List<String> toStringHelper(BinarySearchTreeNode<T> root, ArrayList<String> list, String addedString,
            Stack<String> tmpStack) {
        list.add(addedString);
        if (root.getLeft() != null) {

            toStringHelper(root.getLeft(), list, String.format(" L:(%s", root.getLeft().getName()),
                    tmpStack);
            list.add(")");
        }

        if (root.getRight() != null) {

            toStringHelper(root.getRight(), list, String.format(" R:(%s", root.getRight().getName()), tmpStack);
            list.add(")");
        }

        return list;

    }

    public static void main(String... args) {
        // each tree has a name, provided to its constructor
        BinarySearchTree<Integer> t1 = new BinarySearchTree<>("Oak");
        // adds the elements to t1 in the order 5, 3, 0, and then 9
        t1.addAll(Arrays.asList(5, 3, 0, 9));
        BinarySearchTree<Integer> t2 = new BinarySearchTree<>("Maple");
        // adds the elements to t2 in the order 9, 5, and then 10
        t2.addAll(Arrays.asList(9, 5, 10));
        System.out.println(t1); // see the expected output for exact format
        t1.forEach(System.out::println); // iteration in increasing order
        System.out.println(t2); // see the expected output for exact format
        t2.forEach(System.out::println); // iteration in increasing order
        BinarySearchTree<String> t3 = new BinarySearchTree<>("Cornucopia");
        t3.addAll(Arrays.asList("coconut", "apple", "banana", "plum", "durian",
                "no durians on this tree!", "tamarind"));
        System.out.println(t3); // see the expected output for exact format

        t3.forEach(System.out::println); // iteration in increasing order
        BinarySearchTree<String> t4 = new BinarySearchTree<>("Cornucopia");
        t4.addAll(Arrays.asList("apple"));
        List<String> mergedList = merge(t3, t4);
        for (String i : mergedList) {
            System.out.print(i + " ");
        }

    }
}

class BinarySearchTreeNode<T> {
    private T name;
    private BinarySearchTreeNode<T> left, right;

    public BinarySearchTreeNode(T name) {
        this.left = null;
        this.right = null;
        this.name = name;

    }

    public T getName() {
        return name;
    }

    public BinarySearchTreeNode<T> getLeft() {
        return left;
    }

    public BinarySearchTreeNode<T> getRight() {
        return right;
    }

    public void setRight(BinarySearchTreeNode<T> right) {
        this.right = right;
    }

    public void setLeft(BinarySearchTreeNode<T> left) {
        this.left = left;
    }

    public void setName(T name) {
        this.name = name;
    }
}

File: Programming-Abstraction/Binary-Search_Tree-Java/bin/binarysearchtree.py
class Merger:
    def merge(t1, t2):
        tmp_list = []
        for i in t1:
            tmp_list.append(i)
        for i in t2:
            tmp_list.append(i)
        tmp_list.sort()
        return tmp_list


class BinarySearchTree:
    def __init__(self, name: str, root: 'Node'):
        self.root = root
        self.cursor = Node()
        self.name = name
        self.size = 0

    def cursor_to_root(self):
        self.cursor = self.root

    def cursor_to_left(self):
        self.cursor = self.cursor.left

    def cursor_to_right(self):
        self.cursor = self.cursor.right

    def add_all(self, *lst):
        for i in lst:
            newNode = Node()
            newNode.name = i
            self.add(newNode)
            self.size += 1

    def add(self, newNode: 'Node'):
        if self.root.name == None:
            self.root = newNode
            self.cursor_to_root()
        else:
            if self.cursor.name > newNode.name:
                if (self.cursor.left == None):
                    self.cursor.left = newNode
                    self.cursor_to_root()
                else:
                    self.cursor_to_left()
                    self.add(newNode)

            if self.cursor.name < newNode.name:
                if (self.cursor.right == None):
                    self.cursor.right = newNode
                    self.cursor_to_root()
                else:
                    self.cursor_to_right()
                    self.add(newNode)

    def __iter__(self):
        return self.tree_as_list(self.root)

    def __str__(self):
        s = ''
        self.cursor_to_root()
        tmp_list = self.str_helper(
            self.root, [], '[{}] {}'.format(self.name, self.root.name))
        for i in tmp_list:
            s += "{}".format(i)
        return s

    def str_helper(self, cur: 'Node', lst=[], added_string=''):
        lst.append(added_string)
        if cur.left != None:
            self.str_helper(cur.left, lst, " L:({}".format(cur.left.name))
            lst.append(')')
        if cur.right != None:
            self.str_helper(cur.right, lst, " R:({}".format(cur.right.name))
            lst.append(')')
        return lst

    def tree_as_list(self, root):
        if root != None:

            for x in self.tree_as_list(root.left):
                yield x
            yield root.name
            for x in self.tree_as_list(root.right):
                yield x


class Node:
    def __init__(self):
        self.name = None
        self.left = None
        self.right = None

    def __iter__(self):
        return self.pre_order()

    def pre_order(self):
        if self != None:
            if self.left != None:
                for x in self.left.pre_order():
                    yield x
            yield self.name
            if self.right != None:
                for x in self.right.pre_order():
                    yield x


if __name__ == "__main__":
    t1 = BinarySearchTree(name="Oak", root=Node())
    t2 = BinarySearchTree(name="Birch", root=Node())
    t3 = BinarySearchTree(name="Cornucopia", root=Node())
    t1.add_all(5, 3, 9, 0)
    t2.add_all(1, 0, 10, 2, 7)
    t3.add_all("coconut", "apple", "banana", "plum", "durian",
               "no durians on this tree!", "tamarind")
    for x in t1.root:
        print(x)
    for x in t2.root:
        print(x)
    print(Merger.merge(t1, t2))
